### PR TITLE
Carry #15975 - Add extra fields based on label and env for gelf/fluentd/json-file/journald log drivers

### LINKED ISSUE #15816: Proposal: Add container/image labels to logging driver output
I'm using the new fluentd logging driver in my environment.  It would be great to have the labels that are attached to my container included in the data that goes to fluentd.

### PR DESCRIPTION
**Pull request by vdemeester**
Carry and Closes #15975, adding docs.
Fix #15816.

🐙

> This adds container env and labels to log context, and use the data in gelf and fluentd log driver to expose more logging data.
> 
> Usage:
> - gelf: `--log-opt gelf-labels=label1,label2 --log-opt gelf-env=env1,env2`
> - fluentd: `--log-opt fluentd-labels=label1,label2 --log-opt fluentd-env=env1,env2`
> 
> The above instructions add `label1`, `label2`, `env1`, `env2` key-value pair to log.

🐸

Putting this in `docs-review` and milestone `1.9.0` as the previous one :wink:.

/ping @thaJeztah @moxiegirl

### CODE REVIEW

**Comment by icecrime** [comment]
Maybe we can also squash then @vdemeester :innocent:

File: docs/reference/logging/overview.md [hunk 1]
```diff
@@ -41,6 +43,21 @@ Logs that reach `max-size` are rolled over. You can set the size in kilobytes(k)
...
+when using this logging driver. Each of these options takes a list of
+key separated by comma. A additional field appears for containers that
+have a environment variable or label matching these keys.
+
+Giving the docker daemon is started with the options
+`--log-driver=json-file --log-opt labels=foo --log-opt env=foo,fizz`,
+running a container with `--label foo=bar -e fizz=buzz` adds the
+`attrs` field that looks like:
+
+    "attrs":{"fizz":"buzz","foo":"bar"}
```
Comment (moxiegirl) [inline docs/reference/logging/overview.md]: Note clear how I would get this to appear `docker inspect` or something else.  I'm assuming `inspect`

File: docs/reference/logging/overview.md [hunk 2]
```diff
@@ -41,6 +43,21 @@ Logs that reach `max-size` are rolled over. You can set the size in kilobytes(k)
 
 If `max-size` and `max-file` are set, `docker logs` only returns the log lines from the newest log file.
 
+The `labels` and `env` options adds additional attributes to be logged
```
Comment (moxiegirl) [inline docs/reference/logging/overview.md]: @vdemeester Some of this isn't clear.  I think examples with actual commands will help. I've taken a stab at it.

--->edit for markdown

The `labels` and `env` options add additional attributes for use with this logging driver. Each of these options takes a comma-separated list of
keys. If there is collision between `label` and `env` keys, the value of the
`env` takes precedence.

To use attributes, specify them when you start the Docker daemon.

```
docker daemon --log-driver=json-file --log-opt labels=foo --log-opt env=foo,fizz
```

Then, run a container and specify values for the `labels` or `env`.  For example, you might use this:

```
docker run --label foo=bar -e fizz=buzz -d -P training/webapp python app.py
````

This adds  field that looks like:

    "attrs":{"fizz":"buzz","foo":"bar"}

Use `docker inspect CONTAINERNAME` to view these values.

File: docs/reference/logging/overview.md (hunk 1 above)
Comment (vdemeester) [inline docs/reference/logging/overview.md]: It's not in `inspect` actually, it appears on the log file (in `/var/lib/docker/container/{ContainerID}/{ContainerID}.json.log`). Had a little trouble to write this part :sweat_smile:.

File: docs/reference/logging/overview.md [hunk 3]
```diff
@@ -128,6 +187,19 @@ If container cannot connect to the Fluentd daemon on the specified address,
 the container stops immediately. For detailed information on working with this
 logging driver, see [the fluentd logging driver](fluentd.md)
 
+The `labels` and `env` options adds additional attributes to be logged
```
Comment (moxiegirl) [inline docs/reference/logging/overview.md]: remove --- no need to repeat for each driver

File: docs/reference/logging/overview.md [hunk 4]
```diff
@@ -112,9 +149,31 @@ By default, Docker uses the first 12 characters of the container ID to tag log m
 Refer to the [log tag option documentation](log_tags.md) for customizing
 the log tag format.
 
+The `labels` and `env` options adds additional attributes to be logged
```
Comment (moxiegirl) [inline docs/reference/logging/overview.md]: remove --- no need to repeat for each driver

File: docs/reference/logging/overview.md [hunk 5]
```diff
@@ -90,16 +107,36 @@ the log tag format.
...
+The journald logging driver supports the following options:
+
+    --log-opt labels=label1,label2
+    --log-opt env=env1,env2
+
 The `journald` logging driver stores the container id in the journal's `CONTAINER_ID` field. For detailed information on
 working with this logging driver, see [the journald logging driver](journald.md)
 reference documentation.
 
+The `labels` and `env` options adds additional attributes to be logged
```
Comment (moxiegirl) [inline docs/reference/logging/overview.md]: remove --- no need to repeat for each driver

File: docs/reference/logging/overview.md (hunk 2 above)
Comment (vdemeester) [inline docs/reference/logging/overview.md]: Yeah definitely better :+1:. I'll update (except for the `docker inspect` part, where I'll probably point out where the log is)

File: docs/reference/logging/overview.md (hunk 1 above)
Comment (moxiegirl) [inline docs/reference/logging/overview.md]: Interesting, I could actually seem them with `inspect` just not in this format.

Comment (vdemeester) [inline docs/reference/logging/overview.md]: Yeah, in inspect it's all of the environment variable and labels. Here it's just the "filtered" one (from the `log-opt`) that ends up in the logs.

File: docs/reference/logging/overview.md (hunk 5 above)
Comment (vdemeester) [inline docs/reference/logging/overview.md]: Well, for each driver the representation differs :

- `json-file` : goes in an `attrs` attribute in the json log file
- `journald` : add log fields to the current logs (something like tags.. I think :sweat_smile:)
- `fluentd` : add entries for each keys in the `extra` attribute
- `gelf` : almost the same as `fluentd`

That's why I repeated it (and the first description is in `json-file`).

I think the first description should go above then (in a `generic` options) and have a small paragraph on the *representation* for each driver that supports it ? @moxiegirl WDYT ?

**Comment by tiborvass** [comment]
@moxiegirl @vdemeester is there agreement here or still debate?

**Comment by vdemeester** [comment]
@moxiegirl Updated :
- Updated the content with yours and removed the duplication of text.
- Kept the `--log-opt` on the other one that supports it to be consistent
- Add a smal sentence/example on the `gelf` one (because the way it's represented is different from `json-file`.
- Update `journald.md` and `fluentd.md` as those are specific documentation for these 2 logging drivers and are detailled there (I missed that at first).

@tiborvass not really far from agreement, pretty sure we can make it for 1.9.0 (and it would be so cool :). :innocent: :baby_chick:

**Comment by moxiegirl** [comment]
Thanks @vdemeester LGTM.

**Comment by thaJeztah** [comment]
LGTM, thanks Europe! :D

**Comment by dqminh** [comment]
Thanks all for carrying this :heart: :fireworks: !

**Comment by KostyaSha** [comment]
Imho it not safe to expose env as usually it contains passwords. imho env is the same thing as configuration file mounted into container, so exposing it to log redundant.

for Labels :+1:

File: docs/reference/logging/journald.md [hunk 6]
```diff
@@ -36,6 +36,31 @@ You can set the logging driver for a specific container by using the
...
+
+```
+docker daemon --log-driver=journald --log-opt labels=foo --log-opt env=foo,fizz
+```
+
+Then, run a container and specify values for the `labels` or `env`.  For example, you might use this:
+
+```
+docker run --label foo=bar -e fizz=buzz -d -P training/webapp python app.py
+````
```
Comment (moxiegirl) [inline docs/reference/logging/journald.md]: ut oh, extra tic

**Comment by vdemeester** [comment]
/ping @albers @sdurrheimer I did not update the completion :sweat:.

In summary, new options are `label` and `env` for `--log-opt` and it's available for the following logging-driver : json-file, journald, fluentd and gelf.

**Comment by vdemeester** [comment]
@KostyaSha It will only *expose* envs and labels that are specified in `log-opt labels=…`, not all of them :wink:.

**Comment by albers** [comment]
Bash completion is added in #17306.

**Comment by sdurrheimer** [comment]
I will do the same for Zsh in the same time I will add completion for log options.

**Comment by jmmills** [comment]
It's my understanding that log-opt labels should be available in the container json-file log, and that this was merged in 1.9.0? 
I'm not seeing my container labels in the log file it self, but perhaps I'm doing this wrong?

    $ (docker version && docker-machine ssh default "sudo -i /bin/sh -c 'cat /var/lib/docker/containers/$(docker run -d --log-opt labels=foobar busybox echo testing)/*.log'" 2>&1) | perl -anE 's/^/    /;print'
    Client:
     Version:      1.8.2
     API version:  1.20
     Go version:   go1.4.2
     Git commit:   0a8c2e3
     Built:        Thu Sep 10 19:10:10 UTC 2015
     OS/Arch:      darwin/amd64
    
    Server:
     Version:      1.9.1
     API version:  1.21
     Go version:   go1.4.3
     Git commit:   a34a1d5
     Built:        Fri Nov 20 17:56:04 UTC 2015
     OS/Arch:      linux/amd64
    Boot2Docker version 1.9.1, build master : cef800b - Fri Nov 20 19:33:59 UTC 2015
    Docker version 1.9.1, build a34a1d5
    {"log":"testing\n","stream":"stdout","time":"2015-12-17T21:30:47.982731443Z"}

I was expecting a label field of some sort in the JSON log file, is this not a correct assumption?

**Comment by dqminh** [comment]
@jmmills your container needs to have a foobar label for `--log-opt labels=foobar` to work

**Comment by uschtwill** [comment]
Sorry, but is this working as advertised?

https://docs.docker.com/engine/admin/logging/overview/

Reading this from the very top:

> The labels and env options add additional attributes for use with logging drivers that accept them. Each option takes a comma-separated list of keys. If there is collision between label and env keys, the value of the env takes precedence.
> 
> To use attributes, specify them when you start the Docker daemon.
> 
> `docker daemon --log-driver=json-file --log-opt labels=foo --log-opt env=foo,fizz`
> Then, run a container and specify values for the labels or env. For example, you might use this:
> 
> `docker run --label foo=bar -e fizz=buzz -d -P training/webapp python app.py`
> This adds additional fields to the log depending on the driver, e.g. for json-file that looks like:
> 
> `"attrs":{"fizz":"buzz","foo":"bar"}`

And this further down in the GELF section:

> The labels and env options are supported by the gelf logging driver. It adds additional key on the extra fields, prefixed by an underscore (_).
> 
> ```
> // […]
> "_foo": "bar",
> "_fizz": "buzz",
> // […]
> ```

I was quite certain, this should be working:

`docker daemon --log-driver=gelf --log-opt labels=testing,moretesting`

```
version: '2'

services:

  logging-test:
    image: hello-world
    labels:
      testing: yesplease
      moretesting: givemelabels
    logging:
      driver: gelf
      options:
        gelf-address: udp://localhost:12201
```

Logstash is receiving the output just fine, but those labels are not appended to the logs as extra fields. What am I missing here?

**Comment by timo-klarshift** [comment]
@uschtwill Late but maybe it helps someone else:

```yml
services.logging-test.logging.options.labels: 'testing,moretesting'
```

You have to tell which labels are exposed to this feature. Imagine it exposing all available labels automatically.

**Comment by uschtwill** [comment]
Thanks @timo-klarshift, better late than never. ;)

**Pull request merged**